pii_recognition/evaluation/character_level_evaluation.py

```diff
@@ -0,0 +1,101 @@
+
+
+def label_encoder(
+    text_length: int, entities: List[Entity], label_to_int: Dict[Any, int],
+) -> List[int]:
+    """Encode entity labels into integers.
+
+    Encode a text at character level according to its entity labels as well as a mapping
+    defined by `label_to_int`. Note multi-tagging is not supported. One entity can have
+    only one label tag.
+
+    Args:
+        text_length: length of a text.
+        entities: entities in a text identified by entity_type, start, end.
+        label_to_int: a mapping between entity labels and integers.
+
+    Returns:
+        Integer code of the text.
+    """
+    if 0 in label_to_int.values():
+        raise ValueError(
+            "Value 0 is reserved! If a character does not belong to any entity, it "
+            "would be assigned with 0."
+        )
+    code = [0] * text_length
+
+    for span in entities:
+        s = span.start
+        e = span.end
+        if e > text_length:
+            raise ValueError(
+                f"Span index is out of range: text length is "
+                f"{text_length} but got span index {e}."
+            )
+        label_name = span.entity_type
+        try:
+            label_code = label_to_int[label_name]
+        except KeyError as err:
+            raise Exception(
+                f"Missing label {str(err)} in 'label_to_int' mapping."
+            )
+
+        code[s:e] = [label_code] * (e - s)
+
+    return code
+
+
+def compute_entity_precisions_for_prediction(
+    text_length: int,
+    true_entities: List[Entity],
+    pred_entities: List[Entity],
+    label_mapping: Dict,
+) -> List[Dict]:
+    """Compute precision for every entity in prediction."""
+    true_code: List[int] = label_encoder(text_length, true_entities, label_mapping)
+
+    precisions = []
+    for pred_entity in pred_entities:
+        pred_entity_code: List[int] = label_encoder(
+            text_length, [pred_entity], label_mapping
+        )
+        int_label: int = label_mapping[pred_entity.entity_type]
+        score = compute_label_precision(true_code, pred_entity_code, int_label)
+
+        span_dict = asdict(pred_entity)
+        span_dict.update({"precision": score})
+        precisions.append(span_dict)
+
+    return precisions
+
+
+def compute_entity_recalls_for_ground_truth(
+    text_length: int,
+    true_entities: List[Entity],
+    pred_entities: List[Entity],
+    label_mapping: Dict,
+) -> List[Dict]:
+    """Compute recall for every entity in ground truth."""
+    pred_code: List[int] = label_encoder(text_length, pred_entities, label_mapping)
+
+    recalls = []
+    for true_entity in true_entities:
+        true_entity_code: List[int] = label_encoder(
+            text_length, [true_entity], label_mapping
+        )
+        int_label: int = label_mapping[true_entity.entity_type]
+        score = compute_label_recall(true_entity_code, pred_code, int_label)
+
+        span_dict = asdict(true_entity)
```

Comment:
i think we should call this entity_dict for naming consistency

Reply:
Good catch. Fixed.